### PR TITLE
Refactor existing policies to follow synchronous workflow

### PR DESCRIPTION
**Pull request by samvaity**
Part 1 (Excludes RetryPolicy work): 

This PR refactors existing policies to add synchronous workflow implementations. 

`RetryPolicy` will be covered in a separate PR.

### CODE REVIEW

**Review by kasobol-msft — COMMENTED** [review]
lgtm. please add test coverage.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpRequestLogger.java [hunk 1]
```diff
@@ -34,4 +34,16 @@ default LogLevel getLogLevel(HttpRequestLoggingContext loggingOptions) {
...
+    /**
+     * Logs the HTTP request.
+     * To get the {@link LogLevel} used to log the HTTP request use {@link #getLogLevel(HttpRequestLoggingContext)}.
+     *
+     * @param logger The {@link ClientLogger} used to log the HTTP request.
+     * @param loggingOptions The information available during request logging.
+     * @throws UnsupportedOperationException If the method is not overridden.
+     */
+    default void logRequestSync(ClientLogger logger, HttpRequestLoggingContext loggingOptions) {
+        throw new UnsupportedOperationException("Must be overridden");
```
Comment (kasobol-msft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpRequestLogger.java:47]: should we block on async by default ?

**Review by kasobol-msft — APPROVED** [review]

File: sdk/core/azure-core/src/test/java/com/azure/core/SyncAsyncExtension.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core;
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/core/azure-core/src/test/java/com/azure/core/SyncAsyncExtension.java:4]: Would adding a package need us to export the extensions package for it to be made available across the test module?

>Unable to make public com.azure.core.extensions.SyncAsyncExtension() accessible: module com.azure.core does not "exports com.azure.core.extensions" to unnamed module

**Review by srnagar — APPROVED** [review]
LGTM!

**Pull request merged**